File: src/data/sessions.ts
export interface SessionRow {
  session: string;
  user: string;
  name: string;
  dept: string;
  mac: string;
  method: '有线' | 'WiFi';
  nas: string;
  nasSub: string;
  nasIp: string;
  nasPort: string;
  called: string;
  ip: string;
  vlan: string; // VLAN ID
  vlanLabel: string; // 如 "20 · 研发"
  auth: string;
  duration: string;
  status: '在线' | '待重认证';
  filterId: string;
  timeout: string;
  start: string;
}

/** 与原型 sessions.html 表格行一一对应 */
export const SESSION_ROWS: SessionRow[] = [
  {
    session: '8F3A-0021-C4', user: 'wang.lei', name: '王磊', dept: '研发中心',
    mac: '3C:52:82:1A:4B:01', method: '有线', nas: 'SW-3F-01', nasSub: 'Gi1/0/12',
    nasIp: '10.99.0.11', nasPort: 'Gi1/0/12', called: '00-1D-45-AC-10-01', ip: '10.20.3.41',
    vlan: '20', vlanLabel: '20 · 研发', auth: 'EAP-TLS', duration: '4h 12m', status: '在线',
    filterId: 'acl_rd_std', timeout: '28800', start: '2026-07-27 06:12:44',
  },
  {
    session: '8F3A-0022-17', user: 'li.na', name: '李娜', dept: '市场部',
    mac: 'A4:83:E7:22:9C:7E', method: 'WiFi', nas: 'AC-HQ-01', nasSub: 'AP-3F-012 · SSID Corp',
    nasIp: '10.99.0.30', nasPort: 'AP-3F-012', called: '84-16-F9-0A-33-20:Corp', ip: '10.10.5.87',
    vlan: '10', vlanLabel: '10 · 办公', auth: 'PEAP-MSCHAPv2', duration: '2h 47m', status: '在线',
    filterId: 'acl_staff', timeout: '28800', start: '2026-07-27 07:37:02',
  },
  {
    session: '8F3A-0023-9B', user: 'zhang.wei', name: '张伟', dept: '财务部',
    mac: '7C:2E:DD:41:0A:93', method: '有线', nas: 'SW-5F-02', nasSub: 'Gi1/0/03',
    nasIp: '10.99.0.13', nasPort: 'Gi1/0/03', called: '00-1D-45-AC-14-01', ip: '10.40.2.19',
    vlan: '40', vlanLabel: '40 · 财务隔离', auth: 'EAP-TLS', duration: '6h 03m', status: '在线',
    filterId: 'acl_fin_iso', timeout: '14400', start: '2026-07-27 04:21:36',
  },
  {
    session: '8F3A-0024-E0', user: 'chen.chen', name: '陈晨', dept: '研发中心',
    mac: '00:1A:2B:3C:4D:5E', method: 'WiFi', nas: 'AC-HQ-01', nasSub: 'AP-4F-007 · SSID Corp',
    nasIp: '10.99.0.30', nasPort: 'AP-4F-007', called: '84-16-F9-0A-41-08:Corp', ip: '10.20.6.52',
    vlan: '20', vlanLabel: '20 · 研发', auth: 'EAP-TLS', duration: '1h 19m', status: '待重认证',
    filterId: 'acl_rd_std', timeout: '28800', start: '2026-07-27 09:05:11',
  },
  {
    session: '8F3A-0025-33', user: 'liu.yang', name: '刘洋', dept: '供应链',
    mac: '8C:85:90:5B:11:2F', method: '有线', nas: 'SW-B1-IDC-01', nasSub: 'Gi1/0/21',
    nasIp: '10.99.0.21', nasPort: 'Gi1/0/21', called: '00-1D-45-B8-02-01', ip: '10.50.1.66',
    vlan: '50', vlanLabel: '50 · 供应链', auth: 'PEAP-MSCHAPv2', duration: '3h 34m', status: '在线',
    filterId: 'acl_scm', timeout: '28800', start: '2026-07-27 06:50:58',
  },
  {
    session: '8F3A-0026-A8', user: 'zhao.min', name: '赵敏', dept: '人事行政',
    mac: 'F4:8C:50:77:BE:09', method: 'WiFi', nas: 'AC-HQ-01', nasSub: 'AP-3F-015 · SSID Corp',
    nasIp: '10.99.0.30', nasPort: 'AP-3F-015', called: '84-16-F9-0A-33-2F:Corp', ip: '10.10.4.120',
    vlan: '10', vlanLabel: '10 · 办公', auth: 'PEAP-MSCHAPv2', duration: '5h 51m', status: '在线',
    filterId: 'acl_staff', timeout: '28800', start: '2026-07-27 04:33:45',
  },
  {
    session: '8F3A-0027-5C', user: 'sun.peng', name: '孙鹏', dept: '研发中心',
    mac: '3C:52:82:1A:8D:40', method: '有线', nas: 'SW-3F-01', nasSub: 'Gi1/0/07',
    nasIp: '10.99.0.11', nasPort: 'Gi1/0/07', called: '00-1D-45-AC-10-01', ip: '10.20.3.15',
    vlan: '20', vlanLabel: '20 · 研发', auth: 'EAP-TLS', duration: '7h 26m', status: '在线',
    filterId: 'acl_rd_std', timeout: '28800', start: '2026-07-27 02:58:20',
  },
  {
    session: '8F3A-0028-71', user: 'zhou.ting', name: '周婷', dept: '市场部',
    mac: 'DC:A6:32:99:04:71', method: 'WiFi', nas: 'AC-HQ-01', nasSub: 'AP-5F-003 · SSID Corp',
    nasIp: '10.99.0.30', nasPort: 'AP-5F-003', called: '84-16-F9-0A-50-13:Corp', ip: '10.10.7.34',
    vlan: '10', vlanLabel: '10 · 办公', auth: 'EAP-FAST', duration: '0h 42m', status: '在线',
    filterId: 'acl_staff', timeout: '28800', start: '2026-07-27 09:42:53',
  },
  {
    session: '8F3A-0029-0D', user: 'wu.hao', name: '吴昊', dept: '运维部',
    mac: '48:4D:7E:C3:56:B8', method: '有线', nas: 'SW-3F-02', nasSub: 'Gi1/0/18',
    nasIp: '10.99.0.12', nasPort: 'Gi1/0/18', called: '00-1D-45-AC-12-01', ip: '10.99.2.8',
    vlan: '99', vlanLabel: '99 · 运维管理', auth: 'EAP-TLS', duration: '9h 15m', status: '在线',
    filterId: 'acl_ops_admin', timeout: '43200', start: '2026-07-27 01:09:37',
  },
  {
    session: '8F3A-0030-B6', user: 'zheng.nan', name: '郑楠', dept: '财务部',
    mac: 'B0:6E:BF:12:78:E3', method: 'WiFi', nas: 'AC-HQ-01', nasSub: 'AP-5F-006 · SSID Corp',
    nasIp: '10.99.0.30', nasPort: 'AP-5F-006', called: '84-16-F9-0A-50-26:Corp', ip: '10.40.3.91',
    vlan: '40', vlanLabel: '40 · 财务隔离', auth: 'EAP-TLS', duration: '2h 08m', status: '在线',
    filterId: 'acl_fin_iso', timeout: '14400', start: '2026-07-27 08:16:29',
  },
];

export const SESSION_FILTER_OPTIONS = {
  dept: ['全部部门', '研发中心', '市场部', '财务部', '供应链', '人事行政', '运维部'],
  method: ['全部', '有线 802.1X', '办公 WiFi'],
  nas: ['全部设备', 'SW-3F-01', 'SW-3F-02', 'SW-5F-02', 'SW-B1-IDC-01', 'AC-HQ-01'],
  vlan: ['全部', '10 · 办公', '20 · 研发', '40 · 财务隔离', '50 · 供应链', '99 · 运维管理'],
  auth: ['全部', 'EAP-TLS', 'PEAP-MSCHAPv2', 'EAP-FAST'],
} as const;

File: backend/src/openredius/services/sessions.py
"""Session service: active radacct ⋈ app metadata (docs/02 SessionRow mapping)."""

from __future__ import annotations

from dataclasses import dataclass
from datetime import UTC, datetime, timedelta
from types import SimpleNamespace
from typing import Any

from sqlalchemy import Select, String, and_, cast, func, or_, select
from sqlalchemy.ext.asyncio import AsyncSession

from openredius.core.errors import ApiError
from openredius.core.listing import PageParams, apply_sort
from openredius.models import AccessUser, NasDevice, PolicyGroup, Vlan
from openredius.radius.tables import ip_text, radius_readable, radius_table
from openredius.schemas.sessions import SessionRowOut

_WIRELESS_PREFIX = "Wireless"
_EAP_LABELS = {"eap-tls": "EAP-TLS", "peap-mschapv2": "PEAP-MSCHAPv2"}
_CSV_CAP = 10_000


@dataclass(slots=True)
class SessionJoin:
    """One joined row: flat Core columns + ORM entities."""

    acct: SimpleNamespace
    user: AccessUser | None
    policy: PolicyGroup | None
    vlan: Vlan | None
    nas: NasDevice | None


def _split_row(row: Any, acct_columns: list[str]) -> SessionJoin:
    mapping = row._mapping
    acct = SimpleNamespace(**{c: mapping[c] for c in acct_columns})
    return SessionJoin(
        acct=acct,
        user=row.AccessUser,
        policy=row.PolicyGroup,
        vlan=row.Vlan,
        nas=row.NasDevice,
    )


@dataclass(slots=True)
class SessionFilters:
    dept: str | None = None
    method: str | None = None  # 有线 / WiFi / wired / wifi
    nas: str | None = None
    vlan: str | None = None
    auth: str | None = None
    q: str | None = None


def _radacct(db: AsyncSession):
    return radius_table(db.get_bind().dialect.name, "radacct")


def _normalize_method(value: str | None) -> str | None:
    if not value:
        return None
    v = value.strip().lower()
    if v in {"wifi", "无线", "wireless"}:
        return "wifi"
    if v in {"有线", "wired", "ethernet"}:
        return "wired"
    raise ApiError("invalid_method", f"unsupported method filter: {value}", 422)


def _normalize_auth(value: str | None) -> str | None:
    if not value:
        return None
    for key, label in _EAP_LABELS.items():
        if value.strip().lower() in {key, label.lower()}:
            return key
    raise ApiError("invalid_auth", f"unsupported auth filter: {value}", 422)


def _base_query(db: AsyncSession, active_only: bool, f: SessionFilters) -> Select:
    radacct = _radacct(db)
    # radacct.nasipaddress is `inet` on PostgreSQL — host() for string ops/joins.
    nas_ip = ip_text(db, radacct.c.nasipaddress)
    stmt = (
        select(radacct, AccessUser, PolicyGroup, Vlan, NasDevice)
        .select_from(radacct)
        .join(AccessUser, AccessUser.account == radacct.c.username, isouter=True)
        .join(PolicyGroup, PolicyGroup.id == AccessUser.policy_group_id, isouter=True)
        .join(Vlan, Vlan.id == PolicyGroup.vlan_id, isouter=True)
        .join(NasDevice, NasDevice.nasname == nas_ip, isouter=True)
    )
    if active_only:
        stmt = stmt.where(radacct.c.acctstoptime.is_(None))

    if f.dept:
        stmt = stmt.where(AccessUser.dept.ilike(f"%{f.dept}%"))
    method = _normalize_method(f.method)
    if method == "wifi":
        stmt = stmt.where(radacct.c.nasporttype.like(f"{_WIRELESS_PREFIX}%"))
    elif method == "wired":
        stmt = stmt.where(
            or_(
                radacct.c.nasporttype.is_(None),
                radacct.c.nasporttype.not_like(f"{_WIRELESS_PREFIX}%"),
            )
        )
    if f.nas:
        like = f"%{f.nas}%"
        stmt = stmt.where(or_(NasDevice.name.ilike(like), nas_ip.ilike(like)))
    if f.vlan:
        try:
            vid = int(f.vlan)
        except ValueError as exc:
            raise ApiError("invalid_vlan", f"vlan must be an integer: {f.vlan}", 422) from exc
        stmt = stmt.where(Vlan.vid == vid)
    auth = _normalize_auth(f.auth)
    if auth:
        stmt = stmt.where(PolicyGroup.eap_method == auth)
    if f.q:
        like = f"%{f.q}%"
        framed_ip = ip_text(db, radacct.c.framedipaddress)
        stmt = stmt.where(
            or_(
                radacct.c.username.ilike(like),
                AccessUser.name.ilike(like),
                radacct.c.callingstationid.ilike(like),
                framed_ip.ilike(like),
            )
        )
    return stmt


def _row_out(row: SessionJoin, now: datetime) -> SessionRowOut:
    acct, user, policy, vlan, nas = row.acct, row.user, row.policy, row.vlan, row.nas
    start = acct.acctstarttime.replace(tzinfo=UTC) if acct.acctstarttime else now
    elapsed = max(int((now - start).total_seconds()), 0)
    reauth = policy.reauth_interval_s if policy else None
    status = "reauth-pending" if reauth and elapsed > reauth else "online"
    nasporttype = acct.nasporttype or ""
    method = "WiFi" if nasporttype.startswith(_WIRELESS_PREFIX) else "有线"
    eap = policy.eap_method if policy else None
    vid = vlan.vid if vlan else None
    return SessionRowOut(
        acct_unique_id=acct.acctuniqueid,
        acct_session_id=acct.acctsessionid,
        username=acct.username,
        name=user.name if user else "",
        dept=user.dept if user else "",
        mac=acct.callingstationid or "",
        method=method,
        nas_name=nas.name if nas else "",
        nas_area=nas.area if nas else "",
        nas_ip=str(acct.nasipaddress),
        nas_port=acct.nasportid or "",
        called=acct.calledstationid or "",
        ip=str(acct.framedipaddress or ""),
        vlan=str(vid) if vid is not None else "",
        vlan_label=f"{vid} · {vlan.name}" if vlan else "",
        auth_method=_EAP_LABELS.get(eap, eap or ""),
        duration_s=elapsed,
        status=status,
        filter_id=policy.acl_name if policy else "",
        session_timeout=(
            str(policy.session_timeout_s) if policy and policy.session_timeout_s else ""
        ),
        start=start,
        bytes_in=int(acct.acctinputoctets or 0),
        bytes_out=int(acct.acctoutputoctets or 0),
    )


async def list_sessions(
    db: AsyncSession, f: SessionFilters, params: PageParams, sort: str | None = None
) -> dict[str, Any]:
    if not await radius_readable(db):
        return {"items": [], "total": 0, "page": params.page, "size": params.size}
    stmt = _base_query(db, active_only=True, f=f)
    total = (await db.execute(select(func.count()).select_from(stmt.subquery()))).scalar_one()
    stmt = apply_sort(
        stmt,
        sort,
        {
            "start": _radacct(db).c.acctstarttime,
            "user": _radacct(db).c.username,
        },
        default="-start",
    )
    rows = (await db.execute(stmt.offset(params.offset).limit(params.size))).all()
    now = datetime.now(UTC)
    cols = [c.name for c in _radacct(db).columns]
    return {
        "items": [_row_out(_split_row(r, cols), now) for r in rows],
        "total": total,
        "page": params.page,
        "size": params.size,
    }


async def get_session_detail(
    db: AsyncSession, acct_unique_id: str
) -> tuple[SessionRowOut, dict[str, Any]] | None:
    """(row DTO, raw radacct attributes) or None."""
    if not await radius_readable(db):
        return None
    radacct = _radacct(db)
    stmt = _base_query(db, active_only=False, f=SessionFilters()).where(
        radacct.c.acctuniqueid == acct_unique_id
    )
    row = (await db.execute(stmt)).first()
    if row is None:
        return None
    joined = _split_row(row, [c.name for c in radacct.columns])
    out = _row_out(joined, datetime.now(UTC))
    attrs = {
        k: (v.isoformat() if isinstance(v, datetime) else v) for k, v in vars(joined.acct).items()
    }
    return out, attrs


async def sessions_for_csv(db: AsyncSession, f: SessionFilters) -> list[SessionRowOut]:
    if not await radius_readable(db):
        return []
    stmt = _base_query(db, active_only=True, f=f).order_by(_radacct(db).c.acctstarttime.desc())
    rows = (await db.execute(stmt.limit(_CSV_CAP))).all()
    now = datetime.now(UTC)
    cols = [c.name for c in _radacct(db).columns]
    return [_row_out(_split_row(r, cols), now) for r in rows]


async def fetch_sessions_by_unique_ids(
    db: AsyncSession, acct_unique_ids: list[str]
) -> list[SessionJoin]:
    """Joined active rows for disconnect handling."""
    if not acct_unique_ids:
        return []
    radacct = _radacct(db)
    stmt = _base_query(db, active_only=True, f=SessionFilters()).where(
        radacct.c.acctuniqueid.in_(acct_unique_ids)
    )
    rows = (await db.execute(stmt)).all()
    cols = [c.name for c in radacct.columns]
    return [_split_row(r, cols) for r in rows]


async def close_session_fallback(db: AsyncSession, acct_unique_id: str) -> None:
    """Backend-side stop when the NAS never sent Accounting-Stop (docs/04 CoA)."""
    radacct = _radacct(db)
    await db.execute(
        radacct.update()
        .where(and_(radacct.c.acctuniqueid == acct_unique_id, radacct.c.acctstoptime.is_(None)))
        .values(
            acctstoptime=datetime.now(UTC).replace(tzinfo=None),
            acctterminatecause="Admin-Reset",
            connectinfo_stop="backend-closed",
        )
    )


@dataclass(slots=True)
class NasActivity:
    last_seen: datetime | None
    active_sessions: int


async def nas_activity(db: AsyncSession) -> dict[str, NasActivity]:
    """Per-NAS last-seen + active-session count (docs/02 NAS 在线状态)."""
    if not await radius_readable(db):
        return {}
    dialect = db.get_bind().dialect.name
    acct = radius_table(dialect, "radacct")
    log = radius_table(dialect, "radpostauth")
    acct_ip = ip_text(db, acct.c.nasipaddress)
    log_ip = cast(log.c.nasipaddress, String(64))

    last_seen: dict[str, datetime] = {}
    for ip, ts in (
        await db.execute(select(log_ip, func.max(log.c.authdate)).group_by(log_ip))
    ).all():
        if ip:
            last_seen[ip] = ts
    for ip, ts in (
        await db.execute(select(acct_ip, func.max(acct.c.acctupdatetime)).group_by(acct_ip))
    ).all():
        if ip and (ip not in last_seen or ts > last_seen[ip]):
            last_seen[ip] = ts

    active: dict[str, int] = {}
    for ip, n in (
        await db.execute(
            select(acct_ip, func.count()).where(acct.c.acctstoptime.is_(None)).group_by(acct_ip)
        )
    ).all():
        if ip:
            active[ip] = n

    ips = set(last_seen) | set(active)
    return {
        ip: NasActivity(last_seen=last_seen.get(ip), active_sessions=active.get(ip, 0))
        for ip in ips
    }


def nas_status(
    activity: NasActivity | None,
    window_s: int,
    capacity: int | None,
    high_load_ratio: float = 0.9,
) -> str:
    """online / offline / high-load (docs/02 派生规则)."""
    if activity is None or activity.last_seen is None:
        return "offline"
    now = datetime.now(UTC).replace(tzinfo=None)
    seen = activity.last_seen
    if seen < now - timedelta(seconds=window_s):
        return "offline"
    if capacity and capacity > 0 and activity.active_sessions / capacity >= high_load_ratio:
        return "high-load"
    return "online"


async def nas_ports(db: AsyncSession, nasname: str) -> list[dict]:
    """Active session detail per NAS port (docs/03 ports drawer)."""
    if not await radius_readable(db):
        return []
    radacct = _radacct(db)
    nas_ip_col = ip_text(db, radacct.c.nasipaddress)
    stmt = (
        select(
            radacct.c.nasportid,
            radacct.c.callingstationid,
            radacct.c.username,
            Vlan.vid,
            AccessUser.name,
        )
        .select_from(radacct)
        .join(AccessUser, AccessUser.account == radacct.c.username, isouter=True)
        .join(PolicyGroup, PolicyGroup.id == AccessUser.policy_group_id, isouter=True)
        .join(Vlan, Vlan.id == PolicyGroup.vlan_id, isouter=True)
        .where(radacct.c.acctstoptime.is_(None), nas_ip_col == nasname)
        .order_by(radacct.c.nasportid)
    )
    rows = (await db.execute(stmt)).all()
    return [
        {
            "port": row.nasportid or "",
            "mac": row.callingstationid or "",
            "user": row.name or row.username or "",
            "vlan": str(row.vid) if row.vid else "",
        }
        for row in rows
    ]


async def nas_ssids(db: AsyncSession, nasname: str) -> list[dict]:
    """SSID-level aggregation for wireless NAS (docs/03 ssids drawer)."""
    if not await radius_readable(db):
        return []
    radacct = _radacct(db)
    nas_ip_col = ip_text(db, radacct.c.nasipaddress)
    stmt = (
        select(
            radacct.c.calledstationid,
            func.count().label("count"),
            PolicyGroup.eap_method,
            Vlan.vid,
        )
        .select_from(radacct)
        .join(AccessUser, AccessUser.account == radacct.c.username, isouter=True)
        .join(PolicyGroup, PolicyGroup.id == AccessUser.policy_group_id, isouter=True)
        .join(Vlan, Vlan.id == PolicyGroup.vlan_id, isouter=True)
        .where(radacct.c.acctstoptime.is_(None), nas_ip_col == nasname)
        .group_by(
            radacct.c.calledstationid,
            PolicyGroup.eap_method,
            Vlan.vid,
        )
        .order_by(func.count().desc())
    )
    rows = (await db.execute(stmt)).all()
    return [
        {
            "ssid": row.calledstationid or "(unknown)",
            "auth": _EAP_LABELS.get(row.eap_method, row.eap_method or "-"),
            "count": str(row.count),
            "vlan": str(row.vid) if row.vid else "-",
        }
        for row in rows
    ]
